First Draft (#1)
ucan-wg/ipld-inline-links: README.md

## Dependencies

- [CID]
- [IPLD]

# 0 Abstract

This document presents "IPLD Inline Links", a format for expanding linked graphs "inline". This is helpful when inspecting data as a developer, for passing data to applications that expect tree-structured data, or when you don't know the configuration of the CID of a linked graph.

# 1 Introduction

## 1.1 Motivation

[IPLD] is a format for describing Merkle DAGs with semantics similar to CBOR plus hash linking to other blocks of IPLD. It is an efficient machine encoding, but can be cumbersome to work with in the broader tools ecosystem and developer ergonomics.

### 1.1.1 Containers

IPLD assumes that links will be referenced against a blockstore. The resulting disconnected structures — while efficient for machines — can be cumbersome to work with, debug, inspect, and transmit in non-IPLD aware applications. [Locality of reference] also plays a role: IPLD links are no guaranteed to resolve. Data that is likely to be accessed together being kept inline avoids going out to the blockstore, which may incur a (fallible) network lookup.

> The ecosystem of tools for non-IPLD specific tools (JSON tools, CBOR tools, etc) is much wider than those designed for IPLD in particular. To make use of them today, a developer either needs to write custom code, or find an ad hoc method for reifying an IPLD DAG into some format without links such as JSON. Packaging data for 

Reviewer: I think last sentence got cut off in the middle

### 1.1.2 Codecs

> By referencing a specific [CID] in a link, the encoding and hash algorithm of that block are forced to be decided ahead of time. While common codecs are supported by most implementations, not all applications will support all possible encodings.

Reviewer: ℹ️ I'm also just realizing that not only we get locality of references in this encoding but also guarantee that all of local links are decodable, which we would not have if we allowed configuring link settings.

- The identity CID is frozen by the Merkle root of the encoding structure

## 1.2 Inline Links

IPLD Inline Links solve for the above by adding a standard way to inline data in an IPLD document without altering its hash. Link configuration (hashing algorithm, data encoding) is inherited from the surrounding context, unless the link is present. Explicitly including the CID to provide this configuration is OPTIONAL.

Inline links do not modify the IPLD data model itself, but rather extend the format in a way that can be round-tripped to and from standard IPLD as desired.

# 2 Format

The basic structure of an inline link is similar to how reference links are represented in the most common IPLD codecs: signalling a link by wrapping the payload in a map with the single `"/"` key. By convention, this signals an IPLD-specific extension to a base codec.

## 2.1 Embed Wrapper

An inline link MUST be signalled by wrapping in a map with a `"/"` key.

``` ipldsch
type EmbedWrapper struct {
  embed InlineLink (rename "/")
}
```

## 2.2 Embedded IPLD Payload

The inlined DAG payload MUST contain the inlined DAG. The CID MAY be present or set to `Null`.

``` ipldsch
type Embed struct {

>   link nullable Link

Reviewer: Schema says this is of link type, but then all the examples use strings. Are examples just out of date or is this incorrect ?

I'd also use `&Any` to be more explicit

```suggestion
  link nullable &Any
```

Author: I've gotten corrected both ways on `&Any` and `Link`. I think `type Link = &Any`, so maybe a matter of taste, but sure I like `&Any` instead of an alias 👍

  data          Any
}
```

### 2.2.1 Explicit

``` js
{
  "name": "Alonzo Church",
  "birthday": {
    "/": {

>       "link": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4", // DAG-CBOR & SHA2-256

Reviewer: I think we may be better off just putting a digest, as it would guarantee that the whole DAG can be imported or not at all. If we have CIDs they potentially could use alternative codecs and hashing algorithms and while I may have access to the outer codec / hasher I many not have for the nested ones.

Alternatively spec could say all inline links MUST use same configuration

Reviewer 2: Yeah I was thinking about this as well. The trade-off here being "single source of truth" vs. "easily debuggable, copy-able CIDs".
We rarely see just base-encoded multihashes in the wild & there's very few tools to handle them.

It seems like the spec here is somewhat in-between "tooling" and "a better wire format"...

Author: Yeah, I agree. Your proposal basically cuts a feature that we could add later if we find that it's painful without it — much harder to remove it later than add it in later.

Let's have as small a surface area as possible 👍 I'll make a sub-PR to see what that looks like.

Author: Hmm if we have this, then why even include the hash? We can always calculate it directly from the data, and presumably we want to do that check anyway. I'm going to try a version without a hash _at all_ and let's see how that feels.

Reviewer: > Hmm if we have this, then why even include the hash?

It could allow you incremental verification, but then most IPLD encodings do not support partial or streaming decode so probably it will not be very useful in practice.

>       "link": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4", // DAG-CBOR & SHA2-256

Reviewer: Assuming schema is correct and example isn't

```suggestion
      "link": { "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4" }, // DAG-CBOR & SHA2-256
```

Author: Yep, it doens't need the special `/` (or whatever per codec) because we can infer the type from the schema. We can put it in there, but 1. I'm not sure that really gives us anything, and 2. I'm about to write a PR to test out removing the ability to include explicit CIDs.

Author: Oh, I see what you mean — because it's `Link` 🤔🤔🤔 That's a good point. Anyhow, I'm still doing the PR for point 2 above. Merging your `suggestion` block in the meantime 👍

      "data": {
        "day": 14,
        "month": 6
      }
    }
  }
}
```

Inherited encoding is helpful in many situations. A core intention of IPLD is to abstract away from the exact data encoding. When including a link in IPLD, a hash algorithm and encoding MUST be hardcoded. Such choices are often premature, and lead to awkward design choices to avoid this encoding, or exceptions in prose.

Inheriting from the surrounding context enables a level of flexibility for the nested data. Its hash and encoding MUST be the same as the parent, which is by far the most common case. In order to calculate the hash of the entire structure, the direct parent's encoding MUST be known, then passed to the encoder for the nested graph, the CID calculated, and placed into the parent. See more in [IPLD Canonicalization].

The implicit strategy MAY lead to cases where the same DAG generates different CIDs in the same structure, based on different parent encoding contexts.

## 2.3 Capsule

To signal that a payload contains inlined links, the following capsule MAY be used:

``` ipldsch
type InlineLinkCapsule struct {
  c Any (rename "ipld/inline/v0.1")

> }

Reviewer: If we adopt `{ "/": { ".": ... } }` we probably won't need a separate capsule either, although `.` maybe not descriptive enough.

Author: Yeah agreed. Having an outermost capsule can be nice to signal that this is in our special format, but capsules are always optional.

Included here for completeness, a tabling strategy SHOULD be used when no inlining is desired. This is often the most efficient strategy for storage and retrieval of cross-linked data.

``` mermaid
flowchart LR
    CAR --> a
    a -.->|ref| b

    CAR --> b

    CAR --> c
    b ~~~ c
    a -.->|ref| c
    
    CAR --> d
    b -.->|ref| d
    c -.->|ref| d

    CAR --> e
    d -.->|ref| e
```

# 4 IPLD Canonicalization

Inline links MUST NOT be used in calculation of CIDs. The only time that inline links interact with CIDs MUST be when inheriting the CID configuration from the parent context.

To calculate the CID of a DAG that contains inline links, first walk the graph and replace all inline links with regular CID links (i.e. of the format `{"/": <cid>}`). This MUST be performed recursively when encountering an inline link. Once converted to normal IPLD, proceed as normal.

## 4.1 Example

``` js
// Implicit inline inside DAG-CBOR with SHA2-256
{
  "name": "Alonzo Church",
  "birthday": {
    "/": {
      "link": null,
      "data": {
        "day": 14,
        "month": 6
      }
    }
  }
}

// Converted to explicit CID
{
  "name": "Alonzo Church",
  "birthday": {
    "/": {

>       "link": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4",

Reviewer: Assuming schema is correct and example is out of date 

```suggestion
      "link": { "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4" },
```

While more flexible, the use cases for it seem to be very niche. Given the increased complexity, it was decided against. If more granular configuration is later found to be a critical feature, it can be added at that time.

## 5.2 Why Shallow Nesting?

The two options explored in this design were nesting the link under a `"/"` key, or adding another key as a sibling at the same level. For these reasons given below, it was decided that the nested strategy is the least likely to be misinterpreted or misimplemented.

### 5.2.1 Sibling DAG

``` json
{ 
  "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4",
  ".": {
    "day": 14,
    "month": 6
  }
}

{ 
  "/": null,
  ".": {
    "day": 14,
    "month": 6
  }
}
```

The advantageous features of the sibling strategy include:

- The link (when present) looks exactly like a normal CID
- It saves a few characters in JSON

### 5.2.2 Shallowly Nested

This is the one described in the specification.

``` json
{ 
  "/": {

>     "link": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4",

Reviewer: Assuming schema is correct and example is out of date 

```suggestion
    "link": { "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4" },
```

# IPLD Inline Link Specification v0.1.0

## Editors

- [Brooklyn Zelenka], [Fission]

## Authors

- [Irakli Gozalishvili], [Protocol Labs]
- [Philipp Krüger], [Fission] 
- [Brooklyn Zelenka], [Fission]

## Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119].

## Dependencies

- [CID]
- [IPLD]

# 0 Abstract

This document presents "IPLD Inline Links", a format for expanding linked graphs "inline". This is helpful when inspecting data as a developer, for passing data to applications that expect tree-structured data, or when you don't know the configuration of the CID of a linked graph.

# 1 Introduction

## 1.1 Motivation

[IPLD] is a format for describing Merkle DAGs with semantics similar to CBOR plus hash linking to other blocks of IPLD. It is an efficient machine encoding, but can be cumbersome to work with in the broader tools ecosystem and developer ergonomics.

### 1.1.1 Containers

> IPLD assumes that links will be referenced against a blockstore. The resulting disconnected structures — while efficient for machines — can be cumbersome to work with, debug, inspect, and transmit in non-IPLD aware applications.

Reviewer: I'm not sure how to rephrase this, but if I could I would reframe it as a locality problem, that is IPLD assumes that references can be resolved somehow. In some cases you may have referenced blocks locally (in some blockstore) in other cases you may have to find it in the inter planetary network and sometimes you may fail that.

Which is problematic whenever you need some transactional guarantees.

Author: Great point! Will add.

The ecosystem of tools for non-IPLD specific tools (JSON tools, CBOR tools, etc) is much wider than those designed for IPLD in particular. To make use of them today, a developer either needs to write custom code, or find an ad hoc method for reifying an IPLD DAG into some format without links such as JSON. Packaging data for 

### 1.1.2 Codecs

By referencing a specific [CID] in a link, the encoding and hash algorithm of that block are forced to be decided ahead of time. While common codecs are supported by most implementations, not all applications will support all possible encodings.

The identity hash has been used for similar use cases, but this has several significant drawbacks including:

- A codec and hash algorithm MUST be decided in advance
- The resulting CID will not directly work with non-IPLD tools (at minimum there's a header)
- Depending on the parameters other than hash, the data may be illegible without further processing
- The identity CID is frozen by the Merkle root of the encoding structure

> ## 1.2 Inline Links

Reviewer: I have personally had been calling those embeds to disambiguate from links that use identity hashes which people often refer to as inline links.

Not arguing against "inline link" as a name, just brining up for consideration.

Author: Oh yeah, `embedded-ipld` or `ipld-dag-embed` or similar is also a great name, and makes the structs less wordy! Happy to make that change. @matheus23 thoughts?

> IPLD Inline Links solve for the above by adding a standard way to inline data in an IPLD document without altering its hash. The ability to inherit the CID configuration (hash, algorithm, base encoding, etc) from the surrounding context is included, but is OPTIONAL.

Reviewer: > The ability to inherit the CID configuration (hash, algorithm, base encoding, etc) from the surrounding context is included, but is OPTIONAL

I find this phrasing confusing. I would personally say something along the lines of

```suggestion
IPLD Inline Links solve for the above by adding a standard way to inline data in an IPLD document without altering its hash. Link configuration (hashing algorithm, data encoding) is inherited from the surrounding context, unless link is present, explicitly detailing that configuration.
```

> Inline links do not modify the IPLD data model itself, but rather extend the format in a way that can be round-tripped to and from standard IPLD as desired.

Reviewer: I think it may be easier to frame this as "DAG encoding" format as opposed to IPLD extension which gets confusing, especially around how it fits within current toolchains.

Author: Do you mean something along the lines of this?

```suggestion
Inline links do not extend the IPLD data model. They act as a DAG embedding that can be stripped and returned to regular IPLD.
```

Reviewer: I would say something like

```suggestion
Inline links do not extend the IPLD data model. They define an IPLD DAG serialized format that can be used to store or transfer DAGs.
```

Reviewer: I think a section comparing to CAR format might be a good idea.

# 2 Format

The basic structure of an inline link is similar to how reference links are represented in the most common IPLD codecs: signalling a link by wrapping the payload in a map with the single `"/"` key. By convention, this signals an IPLD-specific extension to a base codec.

## 2.1 Inline Wrapper

An inline link MUST be signalled by wrapping in a map with a `"/"` key.

``` ipldsch
type InlineWrapper struct {
  inline InlineLink (rename "/")
}
```

## 2.2 Inlined DAG Payload

> The inlined DAG payload MUST contain the inlined DAG. The CID MAY be present, omitted, or set to `null`.

Reviewer: Can we limit cardinality to 2 ? Having it optional or `null` seems unnecessary source of variability

Author: haha I knew that this was going to come up 😛 I figured having it optional makes it harder to screw up an implementation and parsing both is very straightforward for most parsers.

I _think_ that I prefer forcing a `null` — any thoughts?

Author: Updated to require the explicit `Null`

``` ipldsch
type InlineLink struct {

>   cid nullable Link (implicit Null)

Reviewer: Most of IPLD spec avoids term “cid” in favor of “link”. Introducing it here seems kind on unfortunate.

Author: Okay will do 👍

Author: Done

>   dag          Any

Reviewer: Since IPLD stands for inter planetary linked **data** perhaps this field should be called “data” ?

The “.” also seems somewhat in place.

Author: `data` makes sense 👍 

> The “.” also seems somewhat in place.

Yeah that was actually the original draft — I had advoctaed for the `{"/": ..., ".": ...}` version in the FAQ back in Brussels because it's so short, but thought better of it for reasons that I explain in other comments. I'm comfortable with this version though, but do read the reasoning elsewhere first.

}

> ```

Reviewer: When we were discussing this at IPFS Thing I was picturing something along the following lines instead:

```js
{
  "name": "Alonzo Church",
  "birthday": {
    "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4", // DAG-CBOR & SHA2-256
    "day": 14,
    "month": 6
  }
}
```

Idea was that `/` could be reserved for the integrity check. I was also hoping we could add support into into existing codecs  that would simply ignore fields other than `/`. New tools aware of the format would additionally perform integrity checks.

Above does not really addressed case where you may not have a CID, which I love that this spec addresses. I wonder if some hybrid module could work perhaps something like:

```js
{
  "name": "Alonzo Church",
  "birthday": {
    "/": {
      ".": {
        // Optional CID for integrity checks or for overriding link encoding / hashing
        "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4",
        "day": 14,
        "month": 6
      }
    }
  }
}
```

My motivation for making CID part of the is following:

- I find myself building trees to be encoded as DAGs
- Sometimes I decode the block and I have a CID
- Sometimes I compute CID lazily for the data

In all of those cases node itself tends to be a natural entity to hold the CID and it is very convenient to say encode some structure into an IR and during encode lazily compute CIDs of nodes that don't already have them as opposed to translating IR into another IR.

In above cases it feels more convenient to have an optional field as opposed to a wrapper, which is why I feel like having cid inside the data itself fits better, although I'm sure there might be good counter arguments to my line of thinking as well.

--

Other thing worth considering is the way `/` is been utilized by DAG-JSON for tagging. Specifically not only `{ '/': string }` is interpreted as link also `{ '/': { bytes: string } }` is interpreted as [bytes](https://ipld.io/specs/codecs/dag-json/spec/#bytes). Adding a `{ '/': { cid: string, dag: any } }` feels somehow inconsistent, while `{ '/': { '.': { ... } } }` fits better IMO.

Author: I think that the fully flat syntax is abiguous / potentially dangerous / has more edge cases:

```
{
  "name": "Alonzo Church",
  "birthday": {
    "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4", // DAG-CBOR & SHA2-256
    "day": 14,
    "month": 6
  }
}
```

If there's a `"/"` key in the payload, it's impossible to tell if it's meant as a link or part of the payload. At minimum separating (nesting) the payload seems like a good idea.

```js
// Modified from above
{
  "name": "Alonzo Church",
  "birthday": {
     // Optional CID for integrity checks or for overriding link encoding / hashing
     "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4",
     ".": {
        "day": 14,
        "month": 6
      }
    }
  }
}
```

I'm comfortable with ☝️ than the one where the CID is mixed with the payload, though see my comment elsewhere on what I find the drawbacks to be.

Reviewer: > I think that the fully flat syntax is abiguous / potentially dangerous / has more edge cases:

I agree with that, however I'm not sure if that applies to proposed `{ "/": { ".": { "/": "bafy...", ... } }` which was attempting to reconcile your points with ones I've mentioned in the comment.

Note that whole thing got pushed under `["/"]["."]` in attempt to disambiguate, perhaps second use of `/` caused a confusion here version with more notes

```js
{
  "name": "Alonzo Church",
  "birthday": {
    //👇 key that DAG-JSON uses to represent json extensions 
    "/": {
    // 👇 key that signals this is embedded block (just like "bytes" in binary) could use "embed" instead 
      ".": {
      // 👇 In this context this is CID for integrity checks
      // we could also use different key if this is confusing overload
        "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4",
        "day": 14,
        "month": 6
      }
    }
  }
}
```

Reviewer: > I'm comfortable with ☝️ than the one where the CID is mixed with the payload, though see my comment elsewhere on what I find the drawbacks to be.

That addresses my points but does introduce drawbacks you've mentioned, which is why attempted to suggest a third option.

Author: Oh, I must have missed that it was `["/"]["."]` when reading the code! Yeah I can live with that version 👍

>
> — [SPJ] & [Simon Marlow], [Secrets of the Glasgow Haskell Compiler inliner][GHC Secrets]

Encoding is trivial on many structures, such as linked lists and trees. The existence of diamond graphs pose a special problem: how many times should a linked graph be inlined versus referenced?

```mermaid
flowchart
    a --> b
    a --> c
    b -.->|ref?| d
    c -.->|ref?| d
    d --> e
```

There are two basic strategies that take advantage of inlining: redundancy and [spanning tree]s. When inlining is not used, the strategy is a form of tabling ([CAR] files and blockstores), and are included here for completeness.

| Representation | Inlining Strategy   | Space | Traversal                      | Typical Implementation      |
|----------------|---------------------|-------|--------------------------------|-----------------------------|
| Redundant Tree | Always              | Large | Fast                           | Standard JSON, CBOR, etc    |

> | Spanning Tree  | Once per unique CID | Small | Often slow; depends on content | DAG-JSON, DAG-CBOR, etc     |

Reviewer 2: > Traversal: Often slow; depends on content

Yeah. My attempt at making this more precise: Lookup & walking a specific path can be slower. Full traversal is fast.

Author: Hmm yeah. Even then it's kind of difficult: if the data is a list or tree, then it's extremely fast. If it's a DAG, then walking the whole graph has a minor penalty where you have to maintain a memoization table. If you walk a path on the DAG that _happens_ to be inlined, then it's extremely fast even without memoization.

> | Table          | Never               | Small | Medium                         | [CAR] file, blockstore, etc |

Reviewer 2: I'd argue "Space" should be "Medium" for the table case. Think of a case with lots of very small blocks (let's say a binary Merkle tree), then you'll end up with roughly double space usage compared to the redundant tree method.

```suggestion
| Table          | Never               | Medium | Medium                         | [CAR] file, blockstore, etc |
```

As a data transfer format, this encoding is often convenient. It eliminates the need for a special decoder and can use standard tools from JSON and CBOR. Spanning trees are often much smaller than redundant trees, and equal to or slightly smaller than an equivalent CAR file.

There is a performance penalty when using a spanning tree directly in memory since some links are references. Following those links requires scanning the entire structure per some rule, such as keeping all inlined links as far to the left as possible. However, this is less efficient than the inlined or tabled strategies. It is RECOMMENDED that if the data be unpacked to a more efficient structure (either lazily at runtime, or eagerly ahead of time) if it is heavily cross-linked.

``` mermaid
flowchart
    a --> b
    a --> c
    b --> d
    c -.->|ref| d
    d --> e
```

## 3.3 Tabling

Included here for completeness, a tabling strategy SHOULD be used when no inlining is desired. This is often the most efficient strategy for storage and retrieval of cross-linked data.

``` mermaid

> flowchart LR
>     car --> a
>     a -.->|ref| b
> 
>     car --> b
> 
>     car --> c
>     b ~~~ c
>     a -.->|ref| c
>     
>     car --> d
>     b -.->|ref| d
>     c -.->|ref| d
> 
>     car --> e
>     d -.->|ref| e

Reviewer 2: Small nit. Should this be upper-cased so it's recognized as an acronym?

```suggestion
flowchart LR
    CAR --> a
    a -.->|ref| b

    CAR --> b

    CAR --> c
    b ~~~ c
    a -.->|ref| c
    
    CAR --> d
    b -.->|ref| d
    c -.->|ref| d

    CAR --> e
    d -.->|ref| e
```

    }
  }
}

// Canonical
{
  "birthday": {"/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4"},
  "name": "Alonzo Church"
}
```

# 5 FAQ

## 5.1 Why Shallow Nesting?

The two options explored in this design were nesting the link under a `"/"` key, or adding another key as a sibling at the same level. For these reasons given below, it was decided that the nested strategy is the least likely to be misinterpreted or misimplemented.

### 5.1.1 Sibling DAG

``` json
{ 
  "/": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4",
  ".": {
    "day": 14,
    "month": 6
  }
}

{ 
  "/": null,
  ".": {
    "day": 14,
    "month": 6
  }
}
```

The advantageous features of the sibling strategy include:

- The link (when present) looks exactly like a normal CID
- It saves a few characters in JSON

### 5.1.2 Shallowly Nested

This is the one described in the specification.

``` json
{ 
  "/": {
    "cid": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4",
    "dag": {
      "day": 14,
      "month": 6
    }
  }
}

{ 
  "/": {
    "cid": "null",
    "dag": {
      "day": 14,
      "month": 6
    }
  }
}
```

The advantages of the nesting strategy include:

> - Distinguishes clearly the case where a `null` CID would otherwise be a parse error

Reviewer: I'm not sure I understand this argument as in how is parse error here is more clear than with sibling dag case ? Are you saying that following is clearly an error

```js
{ 
  "/": {
    "cid": "null",
    "dag": {
      "day": 14,
      "month": 6
    }
  }
}
```

while following is not as clearly an error ?

```js
{ 
  "/": "null",
  ".": {
    "day": 14,
    "month": 6
  }
}

Author: The argumentis something like there being more error cases:

- The parser can accdientally skip the `data` field and only parse out the `/` field
- `{"/": null}` with no DAG should fail, so you have to model inlines/embeds as something further from their presentation in various langauge libraries

`{"/": ..., ".": ...}` was actually my first version, because I didn't want to have to follow as many deeply nested links and it "feels" like an extension to normal CIDs. That punning ended up feeling less clear and more prone to error. But, I figured that presentation would be the biggest part of this conversation, so let's chat about it! Do you have a strong preference?

Reviewer: > The parser can accdientally skip the data field and only parse out the / field

This is arguably a feature IMO, I could pass data into e.g. DAG-JSON codec and it will correctly resolve the block ignoring the data field.

Although it is true that if `data` and `/` mismatch block codec my succeed while, dag codec should fail.

That said, I do think that understand the rational and I think it makes sense.

> - Namespaces the keys, so that the key for the nested DAG is not 

Reviewer: I don't understand what you mean here, could this be rephrased somehow ?

Author: I think I'm missing the rest of the sentence. I seem to do this a lot lately; I must be hitting `D` in normal mode or something.

Author: Fixed

# IPLD Inline Link Specification v0.1.0

## Editors

- [Brooklyn Zelenka], [Fission]

## Authors

- [Philipp Krüger], [Fission] 
- [Brooklyn Zelenka], [Fission]

<!-- 
TODO ask them! Otherwise move to the acknowledgements
* [Irakli Gozalishvili], [Protocol Labs]
-->

## Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119].

## Dependencies

- [CID]
- [IPLD]

# 0 Abstract

This document presents "IPLD Inline Links", a format for expanding linked graphs "inline". This is helpful when inspecting data as a developer, for passing data to applications that expect tree-structured data, or when you don't know the configuration of the CID of a linked graph.

# 1 Introduction

## 1.1 Motivation

> [IPLD] is a format for describing Merkle DAGs with semantics similar to CBOR plus hash linking to other blocks of IPLD. It is an efficient machine encoding, but can be cumbersome to worth with in the broader tools ecosystem and developer ergonomics.

Reviewer: ```suggestion
[IPLD] is a format for describing Merkle DAGs with semantics similar to CBOR plus hash linking to other blocks of IPLD. It is an efficient machine encoding, but can be cumbersome to work with in the broader tools ecosystem and developer ergonomics.
```

Reviewer: It may be worth comparing it to CAR format, which this can be a more compact alternative of.
